src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,374 @@
+// Copyright (C) 2021 Joel Rosdahl and other contributors
+//
+// See doc/AUTHORS.adoc for a complete list of contributors.
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include "RedisStorage.hpp"
+
+#include <AtomicFile.hpp>
+#include <Digest.hpp>
+#include <Logging.hpp>
+#include <UmaskScope.hpp>
+#include <Util.hpp>
+#include <assertions.hpp>
+#include <fmtmacros.hpp>
+#include <util/file_utils.hpp>
+#include <util/string_utils.hpp>
+
+#include <third_party/nonstd/string_view.hpp>
+
+#include <hiredis/hiredis.h>
+
+namespace storage {
+namespace secondary {
+
+static struct timeval
+milliseconds_to_timeval(const std::string& msec)
+{
+  int ms = std::stoi(msec);
+  struct timeval tv;
+  tv.tv_sec = ms / 1000;
+  tv.tv_usec = (ms % 1000) * 1000;
+  return tv;
+}
+
+static std::string
+timeval_to_string(struct timeval tv)
+{
+  return FMT("{:.3f}s", tv.tv_sec + tv.tv_usec / 1000000.0);
+}
+
+static nonstd::optional<struct timeval>
+parse_connect_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("connect-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<struct timeval>
+parse_operation_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("operation-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<std::string>
+parse_username(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("username");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+static nonstd::optional<std::string>
+parse_password(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("password");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+RedisStorage::RedisStorage(const Url& url, const AttributeMap& attributes)
+  : m_url(url),
+    m_connect_timeout(parse_connect_timeout(attributes)),
+    m_operation_timeout(parse_operation_timeout(attributes)),
+    m_username(parse_username(attributes)),
+    m_password(parse_password(attributes))
+{
+  m_prefix = "ccache"; // TODO: attribute
```

Comment:
Ah, that sounds like a good idea.

Reply:
This is most likely optional (but default), it is when sharing the cache server with some other purpose.